adapters/insticator/insticator.go

```diff
@@ -0,0 +1,335 @@
+
+type reqInsticatorExt struct {
+	Caller []insticatorCaller `json:"caller,omitempty"`
+}
+
+type insticatorCaller struct {
+	Name    string `json:"name,omitempty"`
+	Version string `json:"version,omitempty"`
+}
+
+// caller Info used to track Prebid Server
+// as one of the hops in the request to exchange
+var caller = insticatorCaller{"Prebid-Server", "n/a"}
+
+type bidExt struct {
+	Insticator bidInsticatorExt `json:"insticator,omitempty"`
+}
+
+type bidInsticatorExt struct {
+	MediaType string `json:"mediaType,omitempty"`
+}
+
+// Builder builds a new instance of the Insticator adapter with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// getMediaTypeForBid figures out which media type this bid is for
+func getMediaTypeForBid(bid *openrtb2.Bid) openrtb_ext.BidType {
+	switch bid.MType {
+	case openrtb2.MarkupBanner:
+		return openrtb_ext.BidTypeBanner
+	case openrtb2.MarkupVideo:
+		return openrtb_ext.BidTypeVideo
+	default:
+		return openrtb_ext.BidTypeBanner
+	}
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var adapterRequests []*adapters.RequestData
+	var groupedImps = make(map[string][]openrtb2.Imp)
+
+	reqExt, err := makeReqExt(request)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	request.Ext = reqExt
+
+	// Create a deep copy of the request to avoid modifying the original request
+	requestCopy := *request
+	requestCopy.Site = request.Site
+	requestCopy.App = request.App
+
+	for i := 0; i < len(request.Imp); i++ {
+		impCopy, impKey, err := makeImps(request.Imp[i])
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		// Populate site.publisher.id from imp extension
+		if requestCopy.Site != nil || requestCopy.App != nil {
+			if err := populatePublisherId(&impCopy, &requestCopy); err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		}
+
+		resolvedBidFloor, errFloor := resolveBidFloor(impCopy.BidFloor, impCopy.BidFloorCur, requestInfo)
+		if errFloor != nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: fmt.Sprintf("Error in converting the provided bid floor currency from %s to USD",
+					impCopy.BidFloorCur),
+			})
+			continue
+		} else {
+			if resolvedBidFloor > 0 {
+				impCopy.BidFloor = resolvedBidFloor
+				impCopy.BidFloorCur = "USD"
+			}
+		}
+
+		groupedImps[impKey] = append(groupedImps[impKey], impCopy)
+	}
+
+	for _, impList := range groupedImps {
+		if adapterReq, err := a.makeRequest(requestCopy, impList); err == nil {
+			adapterRequests = append(adapterRequests, adapterReq)
+		} else {
+			errs = append(errs, err)
+		}
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request openrtb2.BidRequest, impList []openrtb2.Imp) (*adapters.RequestData, error) {
```

Comment:
FYI: Since request is a value instead of a pointer, this will implicitly create a shallow copy. I suspect that's intentional since this is creating multiple requests in a loop. Calling out for other reviewers.

Reply:
@SyntaxNode Since, I am using the shallow copy  of request in `MakeRequests` I thought to continue using it for the rest flow as well.

I can assign ` request = &requestCopy` again before calling the `makeRequest` function, let me know what do you suggest?

Reply:
If you'd like to reuse the shallow copy of request from `MakeRequests` here in `makeRequest` you need to change the argument signature to `request *openrtb2.BidRequest` to pass the pointer.

Reply:
Sure @SyntaxNode, I have updated it.

Please let me know if everything looks good now.
Thanks a lot :)